(PE-14996) Add hiera type answer/config generation

highb opened the pull request:
In order to install/upgrade with the new installer, we must generate new hiera hocon `pe.conf`-style configs. This PR will allow the generation of that format if beaker-answers is given the new `type` param with a value of `:hiera`

puppetlabs-jenkins commented:
Can one of the admins verify this patch?

ericwilliamson commented on lib/beaker-answers/answers.rb:
```diff
@@ -40,6 +40,17 @@ class Answers
+        'activity_database_password'     => DEFAULT_ANSWERS[:q_activity_database_password],
+        'rbac_database_password'         => DEFAULT_ANSWERS[:q_rbac_database_password],
+        'orchestrator_database_password' => DEFAULT_ANSWERS[:q_orchestrator_database_password],
+        'use_application_services'       => true,
```
Hmm..So do we want to continue using non-standard database name and users? I think we need to, given it's caught a few issues in the past.

highb replied:
Yeah I was thinking about that but apparently forgot to do it. It'd be good to continue to test that, and I think some of the tests might even be dependent on those names/users. I'll patch that up.

highb replied:
Would those names/users would persist even if we are using cert auth? I'm thinking yes, but could be wrong?

ericwilliamson replied:
> Would those names/users would persist even if we are using cert auth?

They would - which is whats giving me so much trouble making cert auth work with the legacy bash script 😢

highb replied:
:/

highb commented:
If it's that much trouble, shouldn't we just push ahead with meep? :/

highb commented on lib/beaker-answers/answers.rb:
```diff
@@ -25,8 +25,8 @@ class Answers
       :q_puppetdb_port                               => 8081,
+      :q_classifier_database_name                    => 'pe-classifier',
       :q_classifier_database_user                    => 'DFGhjlkj',
-      :q_database_name                               => 'pe-classifier',
```
This was probably a typo. Fixed here, and updated the reference in the 3.4 version file.

highb commented:
@ericwilliamson Any other comments on this?

ericwilliamson commented:
@highb I haven't had a chance to really dig into it / test it. At a glance, the only thing i'm really questioning is the addition of `flatten_keys_to_joined_string` - why is it needed?

ericwilliamson commented on lib/beaker-answers/versions/version20162.rb:
```diff
@@ -26,7 +28,55 @@ def generate_answers
+
+        # We only need to specify passwords if we are using password auth
+        unless @options[:database_cert_auth]
+          hiera_hash["puppet_enterprise::puppetdb_database_password"] = answer_for(@options, "puppet_enterprise::puppetdb_database_password")
```
you've added the passwords at the top level `HIERA_DEFAULT_ANSWERS`, so regardless of flag, will always be included

highb replied:
I have specs verifying those keys are not included when `database_cert_auth` is set, so I don't think they are?

ericwilliamson replied:
oh wow. I see the way the logic flows now...

highb replied:
The `DEFAULT_ANSWERS` hashes are only called if we use the `answer_for` method from `BeakerAnswers::Answers`.

highb commented:
That's so when someone provides an answers hash, they can either specify overrides with `::` delimited keys, like `{ 'puppet_enterprise::console_host' => 'console.vm' }` or via a nested hash like `{ 'puppet_enterprise' => { 'console_host' => 'console.vm' } }` I find the second form much easier to deal with if you are overriding multiple values.

jpartlow commented on lib/beaker-answers/versions/version20162.rb:
```diff
@@ -26,7 +28,55 @@ def generate_answers
+
+        return hiera_hash
+      else
+        raise "Don't know how to generate answers for format #{@format}"
```
Is there an 'format' instance variable?

highb replied:
Oh, that's me changing the variable name everywhere but here. Will fix that up.

jpartlow commented on lib/beaker-answers/answers.rb:
```diff
@@ -55,32 +76,42 @@ def self.supported_versions
     # @return [Hash] A hash (keyed from hosts) containing hashes of answer file
     #   data.
-    def self.create version, hosts, options
+    def self.create(version, hosts, options, type = :bash)
```
Options is a Hash, type is a param being added to the call sig.  The comment above is listing it as an @option element of the options Hash.  That might (?) be the better way to pass this though, rather than adding a parameter?  Although ultimately the type should be based on which version class is selected.  However we do currently need a hybrid class than can work either way.  Perhaps rather than a parameter, set @type = :bash in the Answers#initialize() and override from options; still pondering this.

highb replied:
We do typically interface with the Answers lib through that options hash, so it does seem like a good idea to specify the type there. On versions > 20162 we could just ignore that value?

jpartlow commented on lib/beaker-answers/answers.rb:
```diff
@@ -118,6 +153,24 @@ def answers
+    #  end
+    def answer_hiera(host)
+      # Render pretty JSON, because it is a subset of HOCON
+      json = JSON.pretty_generate(answers[host.name])
```
The answers returned by hiera_answers is not a hash of host => answer pairs (correctly), but this is trying to look up by host.name.  I think answer_hiera needs a spec?

highb replied:
Yeah, that's definitely a gap in specs.

highb replied:
`#answer_string` also isn't spec tested. :/

jpartlow commented on lib/beaker-answers/versions/version20162.rb:
```diff
@@ -26,7 +28,55 @@ def generate_answers
       the_answers[console.name].merge!(orchestrator_db)
 
-      the_answers
+      if @type == :bash
```
This handling feels a little strange.  How about case on type and hand off two two different functions generate_bash_answers and generate_hiera_config

highb replied:
Yeah, that would be cleaner. I'll do that.

highb commented on lib/beaker-answers/answers.rb:
```diff
@@ -55,32 +76,42 @@ def self.supported_versions
     # @param [Array<Beaker::Host>] hosts An array of host objects.
     # @param [Hash] options options for answer files
     # @option options [Symbol] :type Should be one of :upgrade or :install.
+    # @option type [Symbol] :format Should be one of :bash or :hiera
```
D'oh, this should be `type` as well, depending on how our discussion below sorts out.

highb commented:
@jpartlow OK, I think using the `options` hash with a `:format` key makes more sense than trying to add a `type` parameter (which is confusing because we already are using the word `type` to denote `:install` vs `:upgrade`.) I've updated the PR to reflect that line of thought.

I've also added tests for/fixed `answers_hiera` as well as `answers_string` to the 20162 specs, and changed the check for which `@format` to a case instead of a bunch of if/else branching.

Thoughts? @ericwilliamson How does this mesh with your ideas?

highb commented:
Rebased and re-ran specs, found an issue due to an updated key name in #15, so fixed that and pushed it up.

highb commented:
@jpartlow I'm 👍  on your recent changes. Anything else we need to finish up here?

jpartlow commented:
@highb yes, integrated testing in puppetlabs/pe_acceptance_test#908 has found several small problems that I'm trying to work through.

highb commented:
@jpartlow Ah, reading the commit messages, those changes also make sense. How it is faring with https://github.com/puppetlabs/pe_acceptance_tests/pull/908 now?

jpartlow commented:
@highb a meep install is passing; need to test meep upgrade; legacy install/upgrade are up for grabs in that pr.

@highb @ericwilliamson @cthorn42 last commit might warrant some discussion?  It stops setting user/database names by default for every install with meep.  My thinking is that we reserve that testing for automated external postgres testing, where we also test password, and leave the baseline testing as close to what the majority of users would test as possible.

---
(PE-14996) Guard addition of database_name, database_user by upgrade …
BeakerAnswers historically sets database user/name to non default
values.  Typically a user will not do this, and it complicates our
default workflow and subsequent upgrades.  For meep, with this commit,
I'm ignoring the database name/user settings if we're not upgrading.
The reason they are needed on upgrade (currently only from pre-meep
versions), is that a legacy install laid down by Beaker will have these
database settings, and we have to have them in pe.conf when upgrading,
or database access fails.

When upgrading from a meep install (2016.2.1+), a pe.conf should not be
created, and beaker-pe should just call the installer-shim with no
pe.conf file. https://tickets.puppetlabs.com/browse/PE-15351

highb commented:
@jpartlow That line of reasoning makes sense to me, since I can't imagine most users really care what our internal postgres db name/user is. In contrast, as you mentioned, those using an external postgres may have no control over those details so we should definitely test that.

jpartlow commented:
We will merge this and puppetlabs/beaker-pe#5 once we've had some time to test CI with the puppetlabs/pe_acceptance_tests#908 changes.

jpartlow commented on lib/beaker-answers/versions/version20162.rb:
```diff
@@ -24,9 +36,121 @@ def generate_answers
+      end
+
+      # We only need to specify passwords if we are using password auth
+      unless @options[:database_cert_auth]
```
@ericwilliamson where is this getting set?

jpartlow commented:
These changes are working for us in the integration CI.
